### PR TITLE
Task ys/j257 episodes pagination functionality

### PR DESCRIPTION
**Pull request by yaroslav207**

### CODE REVIEW

File: backend/src/data/repositories/episode/episode.repository.ts [hunk 1]
```diff
@@ -3,7 +3,9 @@ import {
   Episode as TEpisode,
   EpisodeCreateDTOPayload,
   EpisodeEditDTOPayload,
+  LoadEpisodesByPodcastIdPayload, PodcastLoadFilter,
 } from '~/common/types/types';
+import {PodcastType} from "../../../../../shared/build";
```
Comment (what1s1ove) [inline backend/src/data/repositories/episode/episode.repository.ts]: 🤔

File: backend/src/services/episode/episode.service.ts [hunk 2]
```diff
@@ -13,6 +13,7 @@ import {
 } from '~/data/repositories/repositories';
 import { shownote, comment, record, image } from '~/services/services';
 import { HttpError } from '~/exceptions/exceptions';
+import {LoadEpisodesByPodcastIdPayload} from "../../../../shared/build/common/types/episode/load-episodes-by-podcast-id-payload.type";
```
Comment (what1s1ove) [inline backend/src/services/episode/episode.service.ts]: 🤔

File: frontend/src/components/podcast/components/pagination/pagination.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
+import { PageSizes } from './common/constants/constants';
```
Comment (what1s1ove) [inline frontend/src/components/podcast/components/pagination/pagination.tsx]: add pagination for the `table component`

File: frontend/src/components/podcast/components/pagination/common/constants/page-sizes.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+const PageSizes = [5, 10, 20];
```
Comment (what1s1ove) [inline frontend/src/components/podcast/components/pagination/common/constants/page-sizes.ts]:
```suggestion
const PAGE_SIZES = [5, 10, 20];
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/components/podcast/healper/get-filter-episode/get-filter-episode.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { EpisodeLoadFilter } from 'common/types/types';
+
+export const getFilterEpisode = (page: number, row: number): EpisodeLoadFilter => {
+  return {
+    offset: (page - 1) * row,
```
Comment (EugeniusGR) [inline frontend/src/components/podcast/healper/get-filter-episode/get-filter-episode.ts]: mb, go const?

File: frontend/src/components/podcast/components/components.ts [hunk 6]
```diff
@@ -1 +1,2 @@
 export { default as EpisodeTable } from './episode-table/episode-table';
+export { default as Pagination } from './pagination/pagination';
```
Comment (what1s1ove) [inline frontend/src/components/podcast/components/components.ts]: this should be a part of `table` common component

File: frontend/src/components/podcast/components/pagination/pagination.tsx [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  const handlePreviousPage = (): void => {
+    setPage(pageIndex - 1);
+  };
+
+  const handleNextPage = (): void => {
+    setPage(pageIndex + 1);
+  };
+
+  const changePages = (e: React.ChangeEvent<HTMLInputElement>): void => {
+    const newPage = Number(e.target.value);
```
Comment (what1s1ove) [inline frontend/src/components/podcast/components/pagination/pagination.tsx]:
```suggestion
  const changePages = (evt: React.ChangeEvent<HTMLInputElement>): void => {
    const newPage = Number(evt.target.value);
```

File: frontend/src/components/podcast/components/pagination/pagination.tsx [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  setRow,
+  setPage,
+  pageIndex,
+  totalCountEpisodes,
+}) => {
+
+  const countPage = Math.ceil(totalCountEpisodes / pageSize);
+
+  const handlePreviousPage = (): void => {
+    setPage(pageIndex - 1);
```
Comment (what1s1ove) [inline frontend/src/components/podcast/components/pagination/pagination.tsx]: add constants for all magic numbers

File: frontend/src/components/podcast/components/pagination/pagination.tsx [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    const newPage = Number(e.target.value);
+
+    if(!newPage || newPage === pageIndex || newPage > countPage || newPage < 1){
+      return;
+    }
+
+    setPage(newPage);
+  };
+
+  const changePageSizes = (e: React.ChangeEvent<HTMLSelectElement>): void => {
```
Comment (what1s1ove) [inline frontend/src/components/podcast/components/pagination/pagination.tsx]:
```suggestion
  const handlePageSizesChange = (e: React.ChangeEvent<HTMLSelectElement>): void => {
```

File: frontend/src/components/podcast/components/pagination/pagination.tsx [hunk 10]
```diff
@@ -0,0 +1,82 @@
+import { PAGE_SIZE } from './common/constants/constants';
+import styles from './styles.module.scss';
+
+type Props = {
+  setRow: (row: number) => void;
+  setPage: (page: number) => void;
```
Comment (what1s1ove) [inline frontend/src/components/podcast/components/pagination/pagination.tsx]: check **A9** criteria

File: frontend/src/components/podcast/components/pagination/pagination.tsx [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    setRow(Number(e.target.value));
+  };
+
+  const canPreviousPage = pageIndex !== 1;
+  const canNextPage = pageIndex !== countPage;
+
+  return (
+    <div className={styles.pagination}>
+      <button className={styles.btnPrev} onClick={handlePreviousPage} disabled={!canPreviousPage}>
+        PREV
```
Comment (what1s1ove) [inline frontend/src/components/podcast/components/pagination/pagination.tsx]: do uppercase with styles

File: frontend/src/components/podcast/components/pagination/pagination.tsx [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+        className={styles.input}
+      >
+        {PAGE_SIZE.map((pageSize): JSX.Element => (
+          <option key={pageSize} value={pageSize}>
+            Row {pageSize}
+          </option>
+        ))}
+      </select>
+      <button className={styles.btnNext} onClick={handleNextPage} disabled={!canNextPage}>
+        NEXT
```
Comment (what1s1ove) [inline frontend/src/components/podcast/components/pagination/pagination.tsx]: do uppercase with styles

File: frontend/src/components/podcast/healper/get-filter-episode/get-filter-episode.ts (hunk 5 above)
Comment (what1s1ove) [reply frontend/src/components/podcast/healper/get-filter-episode/get-filter-episode.ts]: agreed

File: frontend/src/components/podcast/healper/get-filter-episode/get-filter-episode.ts [hunk 13]
```diff
@@ -0,0 +1,8 @@
+import { EpisodeLoadFilter } from 'common/types/types';
+
+export const getFilterEpisode = (page: number, row: number): EpisodeLoadFilter => {
```
Comment (what1s1ove) [inline frontend/src/components/podcast/healper/get-filter-episode/get-filter-episode.ts]: export only at the bottom

File: frontend/src/components/podcast/common/constatnts/constants.ts [hunk 14]
```diff
@@ -0,0 +1,4 @@
+export const DEFAULT_EPISODE_PAGINATION = {
```
Comment (what1s1ove) [inline frontend/src/components/podcast/common/constatnts/constants.ts]: export only at the bottom

File: frontend/src/components/podcast/podcast.tsx [hunk 15]
```diff
@@ -78,6 +90,20 @@ const Podcast: React.FC = () => {
     }
   };
 
+  const setRowEpisodeFilter = (row: number): void => {
+    setEpisodePagination({
+      page: 1,
```
Comment (what1s1ove) [inline frontend/src/components/podcast/podcast.tsx]: add constants for all magic numbers

File: frontend/src/components/podcast/podcast.tsx [hunk 16]
```diff
@@ -209,7 +235,16 @@ const Podcast: React.FC = () => {
             </div>
           </div>
           {episodes.length ? (
-            <EpisodeTable episodes={episodes} />
+            <>
+              {isEpisodesLoading ? <Loader/>  : <EpisodeTable episodes={episodes} /> }
+              <Pagination
```
Comment (what1s1ove) [inline frontend/src/components/podcast/podcast.tsx]: pagination should be a part of `table component`

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/components/common/table/components/pagination/pagination.tsx [hunk 17]
```diff
@@ -0,0 +1,82 @@
+import { PAGE_SIZE, PAGINATION_START_VALUE, PAGINATION_STEP } from './common/constants/constants';
+import styles from './styles.module.scss';
+
+type Props = {
+  handleSetRow: (row: number) => void;
```
Comment (what1s1ove) [inline frontend/src/components/common/table/components/pagination/pagination.tsx]: check **A8** criteria

File: frontend/src/components/common/table/components/pagination/pagination.tsx [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+
+  const handlePreviousPage = (): void => {
+    handleSetPage(pageIndex - PAGINATION_STEP);
+  };
+
+  const handleNextPage = (): void => {
+    handleSetPage(pageIndex + PAGINATION_STEP);
+  };
+
+  const changePages = (evt: React.ChangeEvent<HTMLInputElement>): void => {
```
Comment (what1s1ove) [inline frontend/src/components/common/table/components/pagination/pagination.tsx]: check **A9** criteria

File: frontend/src/components/common/table/table.tsx [hunk 19]
```diff
@@ -2,59 +2,80 @@ import { useTable } from 'hooks/hooks';
...
+import { Pagination } from './components/components';
 
 type Props = {
   columns: Column[];
   data: unknown[];
+  handleSetRow: (row: number) => void;
+  handleSetPage: (page: number) => void;
+  pageIndex: number;
+  pageSize: number;
+  totalCountEpisodes: number;
```
Comment (what1s1ove) [inline frontend/src/components/common/table/table.tsx]: check **A8** criteria
add new props `hasPagination` and make all these props optionals

File: frontend/src/components/podcast/healper/get-filter-episode/get-filter-episode.ts [hunk 20]
```diff
@@ -0,0 +1,11 @@
+import { EpisodeLoadFilter } from 'common/types/types';
```
Comment (what1s1ove) [inline frontend/src/components/podcast/healper/get-filter-episode/get-filter-episode.ts]: frontend/src/components/podcast/healper/get-filter-episode/get-filter-episode.helper.ts

File: frontend/src/components/podcast/podcast.tsx [hunk 21]
```diff
@@ -78,6 +90,20 @@ const Podcast: React.FC = () => {
     }
   };
 
+  const setRowEpisodeFilter = (row: number): void => {
```
Comment (what1s1ove) [inline frontend/src/components/podcast/podcast.tsx]: check **A9** criteria

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/components/podcast/healper/healper.ts [hunk 22]
```diff
@@ -0,0 +1 @@
+export { getFilterEpisode } from './get-filter-episode/get-filter-episode.healper';
```
Comment (what1s1ove) [inline frontend/src/components/podcast/healper/healper.ts]:
```suggestion
export { getFilterEpisode } from './get-filter-episode/get-filter-episode.helper';
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Review by what1s1ove — APPROVED** [review]

**Pull request merged**